aes.go

```diff
@@ -0,0 +1,64 @@
+package hdfs
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"encoding/binary"
+	"fmt"
+)
+
+const (
+	// in FileWriter we use chunks upto aesChunkSize bytes to encrypt data
+	aesChunkSize = 1024 * 1024
+)
+
+// calculateIV `shifts` IV to given offset
+// based on calculateIV from AesCtrCryptoCodec.java
+func calculateIV(offset int64, initIV []byte) ([]byte, error) {
+	if len(initIV) != aes.BlockSize {
+		return nil, fmt.Errorf("calculateIV: invalid iv size: %v", len(initIV))
+	}
+
+	counter := offset / aes.BlockSize
+	iv := make([]byte, aes.BlockSize)
+
+	high := binary.BigEndian.Uint64(initIV[:8])
+	low := binary.BigEndian.Uint64(initIV[8:])
+	origLow := low
+
+	low += uint64(counter)
+	if low < origLow { // wrap
+		high += 1
+	}
+
+	binary.BigEndian.PutUint64(iv, high)
+	binary.BigEndian.PutUint64(iv[8:], low)
+
+	return iv, nil
+}
+
+// aesCtrStep perform AES-CTR XOR operation on given byte string.
+// Once encryption and decryption are exactly the same operation for CTR mode,
+// this function can be used to perform both.
+func aesCtrStep(offset int64, enc *TransparentEncryptionInfo, b []byte) ([]byte, error) {
+	iv, err := calculateIV(offset, enc.iv)
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := aes.NewCipher(enc.key)
```

Comment:
I think we want to have this owned by the writer or reader and reuse it (or maybe the block reader?), since it [involves an allocation](https://cs.opensource.google/go/go/+/refs/tags/go1.17.7:src/crypto/aes/cipher.go).

Reply:
cipher stored in `TransparentEncryptionInfo` now.